web/src/unread_ops.ts

```diff
@@ -52,13 +56,190 @@ export function confirm_mark_all_as_read(): void {
+                            count += 1;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    return count;
+}
+
+function getUnfollowedTopicMessagesCount(): number {
+    let count = 0;
+    const all_streams = stream_data.subscribed_stream_ids();
+
+    for (const stream_id of all_streams) {
+        const topics = stream_topic_history.get_recent_topic_names(stream_id);
+        for (const topic of topics) {
+            const is_followed = get_topic_visibility_policy(stream_id, topic) === 3;
+            if (!is_followed) {
+                const all_messages_in_topic = message_util.get_messages_in_topic(stream_id, topic);
+
+                for (const message of all_messages_in_topic) {
+                    if (message.unread) {
+                        count += 1;
+                    }
+                }
+            }
+        }
+    }
+
+    return count;
+}
+
+export function mark_as_read_with_narrow(narrow: NarrowTerm[], modal_id?: string): void {
+    bulk_update_read_flags_for_narrow(narrow, "add", {}, modal_id);
+}
+
+export function mark_selected_options_as_read(modal_id?: string): void {
+    const selected_option = $("#mark_as_read_option").val();
+
+    let is_processing = false;
+    let processing_done = false;
+
+    if (selected_option) {
+        ui_report.loading(
+            $t_html({defaultMessage: "Processing…"}),
+            $("#request-progress-status-banner"),
+        );
+    }
+
+    if (selected_option === "all_messages") {
+        is_processing = true;
+        mark_all_as_read(modal_id);
+        processing_done = true;
+    }
+
+    if (selected_option === "muted_topics") {
+        is_processing = true;
+        const all_streams = stream_data.subscribed_stream_ids();
+
+        for (const stream_id of all_streams) {
+            const topics = stream_topic_history.get_recent_topic_names(stream_id);
+            const stream_name = stream_data.get_stream_name_from_id(stream_id);
+            for (const topic of topics) {
+                const is_muted = get_topic_visibility_policy(stream_id, topic) === 1;
+                if (is_muted) {
+                    bulk_update_read_flags_for_narrow(
+                        [
+                            {operator: "stream", operand: stream_name},
+                            {operator: "topic", operand: topic},
+                        ],
+                        "add",
+                        {},
+                        modal_id,
+                    );
+                }
+            }
+        }
+        processing_done = true;
+    }
+
+    if (selected_option === "topics_not_followed") {
+        is_processing = true;
+        const narrow_options: NarrowTerm[] = [{operator: "is", operand: "unread", negated: false}];
+        narrow_options.push({operator: "is", operand: "followed", negated: true});
+        mark_as_read_with_narrow(narrow_options, modal_id);
+        processing_done = true;
```

Comment:
Please read `unread` library and get the unread messages which are relevant as per the selected option and ask server to mark them as read.

Reply:
I am not able to get the unread topics from streams.

Reply:
I have optimized the implementation.